File: 2023/ruby/day4.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

INPUT = 'inputs/4'

def winning_numbers(line) = line.split('|')[0].split.map(&:to_i)
def card_numbers(line) = line.split('|')[1].split.map(&:to_i)
def initial_card_amount = File.readlines(INPUT).length
def card_points(card) = winning_numbers(card).common_elements(card_numbers(card))

def multiply_cards_points
  sum = 0
  File.foreach(INPUT) do |card|
    winners = card_points card.split(':')[1]
    sum += 2**(winners - 1) if winners.positive?
  end
  sum
end

def accumulate_cards
  sum = initial_card_amount
  card_amounts = Array.new(initial_card_amount, 1)
  File.foreach(INPUT).with_index do |card, index|
    winners = card_points card.split(':')[1]
    repeat = card_amounts[index]
    (index..index + winners).each { |x| card_amounts[x] += repeat }
    sum += winners * repeat
  end
  sum
end

# adding a common_elements method to the array. im only doing this because it's silly.
class Array
  def common_elements(other)
    (self & other).length
  end
end

# part 1 answer
puts multiply_cards_points

# part 2 answer
puts accumulate_cards

File: 2023/rust/src/day4.rs
pub fn main(part: usize) {
    let input = std::fs::read_to_string("inputs/4").unwrap();

    match part {
        0 => part1(input),
        1 => part2(input),
        _ => panic!(),
    }
}

pub fn part2(input: String) {
    let mut cards_amt: Vec<usize> = vec![1; input.lines().count()];
    let mut sum = input.lines().count();
    for (index, line) in input.lines().enumerate() {
        let repeat = cards_amt[index];
        let (winning_numbers, card) = line.split_once(':').unwrap().1.split_once('|').unwrap();
        let winning_numbers: Vec<&str> = winning_numbers.split_whitespace().collect();
        let winners = card
            .split_whitespace()
            .filter(|x| winning_numbers.contains(x))
            .count();
        cards_amt
            .iter_mut()
            .take(index + winners + 1)
            .skip(index)
            .for_each(|x| *x += repeat);
        sum += winners * repeat;
    }
    println!("{sum}");
}

pub fn part1(input: String) {
    let mut sum = 0;
    for line in input.lines() {
        let (winning_numbers, card) = line.split_once(':').unwrap().1.split_once('|').unwrap();
        let winning_numbers: Vec<&str> = winning_numbers.split_whitespace().collect();
        let winners = card
            .split_whitespace()
            .filter(|x| winning_numbers.contains(x))
            .count();
        if winners > 0 {
            sum += 2_u32.pow((winners - 1) as u32);
        }
    }
    println!("{sum}");
}
